test/src/test/java/jenkins/model/queue/ItemDeletionTest.java

```diff
@@ -0,0 +1,238 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.model.queue;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.nullValue;
+
+import hudson.Extension;
+import hudson.model.Executor;
+import hudson.model.Item;
+import hudson.model.ItemGroup;
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.model.Run;
+import hudson.model.TopLevelItem;
+import hudson.model.TopLevelItemDescriptor;
+import hudson.model.listeners.RunListener;
+import hudson.util.StreamTaskListener;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.Objects;
+import java.util.SortedMap;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import jenkins.model.Jenkins;
+import jenkins.model.lazy.LazyBuildMixIn;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class ItemDeletionTest {
+    private static final Logger LOGGER = Logger.getLogger(ItemDeletionTest.class.getName());
+
+    @Rule
+    public BuildWatcher watcher = new BuildWatcher();
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Test
+    public void waitsForAsynchronousExecutions() throws Throwable {
+        var p = r.createProject(AsyncJob.class);
+        var b = (AsyncRun) Jenkins.get().getQueue().schedule2(p, 0).getItem().getFuture().waitForStart();
+        r.waitForMessage("Starting " + b, b);
+        LOGGER.info(() -> "Deleting " + p);
+        p.delete();
+        LOGGER.info(() -> "Deleted " + p);
+        assertThat(r.jenkins.getItemByFullName(p.getFullName()), nullValue());
+        var executables = Stream.of(r.jenkins.getComputers())
+                .flatMap(c -> c.getAllExecutors().stream())
+                .map(Executor::getCurrentExecutable)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+        assertThat(executables, empty());
+    }
+
+    public static class AsyncJob extends Job<AsyncJob, AsyncRun> implements LazyBuildMixIn.LazyLoadingJob<AsyncJob, AsyncRun>, TopLevelItem, Queue.FlyweightTask {
```

Comment:
Quite a mess, but it allows us to test the bug here. Not sure if it seems useful enough to keep it here or if we would prefer to just test against the real Pipeline job and run types in https://github.com/jenkinsci/workflow-job-plugin/pull/468.

I do wonder how much of the boilerplate could be eliminated here though if we introduced some kind of mock job/run types for testing only. Maybe not very much with all of the type parameters. The interesting logic in overridden methods is really just `Run.run`.

Reply:
IMO there is no reason to bother with this; the downstream PR demonstrates the problem clearly, and e.g. https://github.com/jenkinsci/bom/pull/3629 will highlight regressions soon enough. If we want to catch regressions sooner, I think the effort is better put into https://issues.jenkins.io/browse/JENKINS-45047.

Reply:
Yes, fair enough.

Reply:
Removed in [6ec50c7](https://github.com/jenkinsci/jenkins/pull/9790/commits/6ec50c7e34f905073bc9003fd50a651e710e032f).